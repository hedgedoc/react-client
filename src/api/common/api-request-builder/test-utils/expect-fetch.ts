/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { defaultConfig } from '../../default-config'
import { Mock } from 'ts-mockery'

export const expectFetch = (expectedUrl: string, expectedStatusCode: number, expectedOptions: RequestInit): void => {
  global.fetch = jest.fn((fetchUrl: RequestInfo | URL, fetchOptions?: RequestInit): Promise<Response> => {
    expect(fetchUrl).toEqual(expectedUrl)
    expect(fetchOptions).toStrictEqual({
      ...defaultConfig,
      body: undefined,
      headers: new Headers(),
      ...expectedOptions
    })
    return Promise.resolve(
      Mock.of<Response>({
        status: expectedStatusCode
      })
    )
  })
}
