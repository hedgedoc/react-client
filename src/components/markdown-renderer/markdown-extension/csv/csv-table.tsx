/*
 SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)

 SPDX-License-Identifier: AGPL-3.0-only
 */

import React, { useMemo } from 'react'
import { parseCsv } from './csv-parser'

export interface CsvTableProps {
  code: string
  delimiter: string
  showHeader: boolean
  tableRowClassName?: string
  tableColumnClassName?: string
}

export const CsvTable: React.FC<CsvTableProps> = ({
  code,
  delimiter,
  showHeader,
  tableRowClassName,
  tableColumnClassName
}) => {
  const { rowsWithColumns, headerRow } = useMemo(() => {
    const rowsWithColumns = parseCsv(code.trim(), delimiter)
    let headerRow: string[] = []
    if (showHeader) {
      headerRow = rowsWithColumns.splice(0, 1)[0]
    }
    return { rowsWithColumns, headerRow }
  }, [code, delimiter, showHeader])

  const renderTableHeader = useMemo(
    () =>
      headerRow === [] ? undefined : (
        <thead>
          <tr>
            {headerRow.map((column, columnNumber) => (
              <th key={`header-${columnNumber}`}>{column}</th>
            ))}
          </tr>
        </thead>
      ),
    [headerRow]
  )

  const renderTableBody = useMemo(
    () => (
      <tbody>
        {rowsWithColumns.map((row, rowNumber) => (
          <tr className={tableRowClassName} key={`row-${rowNumber}`}>
            {row.map((column, columnIndex) => (
              <td className={tableColumnClassName} key={`cell-${rowNumber}-${columnIndex}`}>
                {column.replace(/^"|"$/g, '')}
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    ),
    [rowsWithColumns, tableColumnClassName, tableRowClassName]
  )

  return (
    <table className={'csv-html-table table-striped'}>
      {renderTableHeader}
      {renderTableBody}
    </table>
  )
}
