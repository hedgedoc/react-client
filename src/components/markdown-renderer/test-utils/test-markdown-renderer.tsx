/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import React, { useMemo } from 'react'
import { useConvertMarkdownToReactDom } from '../hooks/use-convert-markdown-to-react-dom'
import type { MarkdownExtension } from '../markdown-extension/markdown-extension'
import { StoreProvider } from '../../../redux/store-provider'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownExtension[]
}

export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  const dom = useConvertMarkdownToReactDom(lines, extensions, true, false)

  return <StoreProvider>{dom}</StoreProvider>
}
