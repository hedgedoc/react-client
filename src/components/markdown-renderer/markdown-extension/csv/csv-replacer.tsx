/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import type { Element } from 'domhandler'
import React from 'react'
import { ComponentReplacer } from '../../replace-components/component-replacer'
import { CsvTable } from './csv-table'
import { CodeBlockComponentReplacer } from '../../replace-components/code-block-component-replacer'

/**
 * Detects code blocks with "csv" as language and renders them as table.
 */
export class CsvReplacer extends ComponentReplacer {
  public replace(codeNode: Element): React.ReactElement | undefined {
    const code = CodeBlockComponentReplacer.extractTextFromCodeNode(codeNode, 'csv')
    if (!code) {
      return
    }

    const extraData = codeNode.attribs['data-extra']
    const extraRegex = /\s*(delimiter=([^\s]*))?\s*(header)?/
    const extraInfos = extraRegex.exec(extraData)

    let delimiter = ','
    let showHeader = false

    if (extraInfos) {
      delimiter = extraInfos[2] || delimiter
      showHeader = extraInfos[3] !== undefined
    }

    return <CsvTable code={code} delimiter={delimiter} showHeader={showHeader} />
  }
}
