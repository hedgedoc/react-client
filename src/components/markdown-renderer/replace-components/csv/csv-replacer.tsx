/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import type { Element } from 'domhandler'
import React from 'react'
import { ComponentReplacer } from '../ComponentReplacer'
import { CsvTable } from './csv-table'

/**
 * Detects code blocks with "csv" as language and renders them as table.
 */
export class CsvReplacer extends ComponentReplacer {
  public getReplacement(codeNode: Element): React.ReactElement | undefined {
    if (
      codeNode.name !== 'code' ||
      !codeNode.attribs ||
      !codeNode.attribs['data-highlight-language'] ||
      codeNode.attribs['data-highlight-language'] !== 'csv' ||
      !codeNode.children ||
      !codeNode.children[0]
    ) {
      return
    }

    const code = ComponentReplacer.extractTextChildContent(codeNode)

    const extraData = codeNode.attribs['data-extra']
    const extraRegex = /\s*(delimiter=([^\s]*))?\s*(header)?/
    const extraInfos = extraRegex.exec(extraData)

    let delimiter = ','
    let showHeader = false

    if (extraInfos) {
      delimiter = extraInfos[2] || delimiter
      showHeader = extraInfos[3] !== undefined
    }

    return <CsvTable code={code} delimiter={delimiter} showHeader={showHeader} />
  }
}
