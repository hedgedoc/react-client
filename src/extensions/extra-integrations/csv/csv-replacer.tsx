/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import type { Element } from 'domhandler'
import React from 'react'
import type { NodeReplacement } from '../../../components/markdown-renderer/replace-components/component-replacer'
import { CsvTable } from './csv-table'
import {
  ComponentReplacer,
  DO_NOT_REPLACE
} from '../../../components/markdown-renderer/replace-components/component-replacer'
import { CodeBlockComponentReplacer } from '../../../components/markdown-renderer/replace-components/code-block-component-replacer'

/**
 * Detects code blocks with "csv" as language and renders them as table.
 */
export class CsvReplacer extends ComponentReplacer {
  public replace(codeNode: Element): NodeReplacement {
    const code = CodeBlockComponentReplacer.extractTextFromCodeNode(codeNode, 'csv')
    if (!code) {
      return DO_NOT_REPLACE
    }

    const extraData = codeNode.attribs['data-extra']
    const extraRegex = /\s*(delimiter=([^\s]*))?\s*(header)?/
    const extraInfos = extraRegex.exec(extraData)

    const delimiter = extraInfos?.[2] ?? ','
    const showHeader = extraInfos?.[3] !== undefined

    return <CsvTable code={code} delimiter={delimiter} showHeader={showHeader} />
  }
}
