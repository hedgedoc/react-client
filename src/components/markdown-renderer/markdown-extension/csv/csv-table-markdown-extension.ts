/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { MarkdownExtension } from '../markdown-extension'
import { CsvReplacer } from './csv-replacer'
import type { ComponentReplacer } from '../../replace-components/component-replacer'

/**
 * Adds support for csv tables to the markdown rendering using code fences with "csv" as language.
 */
export class CsvTableMarkdownExtension extends MarkdownExtension {
  public buildReplacers(): ComponentReplacer[] {
    return [new CsvReplacer()]
  }
}
