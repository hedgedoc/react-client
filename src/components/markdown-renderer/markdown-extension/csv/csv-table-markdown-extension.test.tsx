/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import * as CsvTableModule from '../csv/csv-table'
import React from 'react'
import type { CodeProps } from '../../replace-components/code-block-component-replacer'
import { mockI18n } from '../../test-utils/mock-i18n'
import { render } from '@testing-library/react'
import { TestMarkdownRenderer } from '../../test-utils/test-markdown-renderer'
import { CsvTableMarkdownExtension } from './csv-table-markdown-extension'

jest.mock('../csv/csv-table')

describe('CSV Table Markdown Extension', () => {
  beforeAll(async () => {
    jest.spyOn(CsvTableModule, 'CsvTable').mockImplementation((({ code }) => {
      return (
        <span>
          this is a mock for csv frame
          <code>{code}</code>
        </span>
      )
    }) as React.FC<CodeProps>)
    await mockI18n()
  })

  afterAll(() => {
    jest.resetModules()
    jest.restoreAllMocks()
  })

  it('renders a csv codeblock', () => {
    const view = render(
      <TestMarkdownRenderer extensions={[new CsvTableMarkdownExtension()]} content={'```csv\na;b;c\nd;e;f\n```'} />
    )
    expect(view.container).toMatchSnapshot()
  })
})
