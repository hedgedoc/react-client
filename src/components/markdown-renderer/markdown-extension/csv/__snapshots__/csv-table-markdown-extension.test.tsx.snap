// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV Table Markdown Extension renders an abc codeblock 1`] = `
<div>
  <pre>
    <code
      class="csv"
    >
      a;b;c
d;e;f

    </code>
  </pre>
  

</div>
`;
